src/coreComponents/physicsSolvers/fluidFlow/kernels/compositional/zFormulation/PotGradZFormulation.hpp

```diff
@@ -0,0 +1,178 @@
+#include "mesh/ElementRegionManager.hpp"
+
+
+namespace geos
+{
+
+namespace isothermalCompositionalMultiphaseFVMKernelUtilities
+{
+
+template< typename VIEWTYPE >
+using ElementViewConst = ElementRegionManager::ElementViewConst< VIEWTYPE >;
+
+using Deriv = constitutive::multifluid::DerivativeOffset;
+
+struct PotGradZFormulation
+{
+  template< integer numComp, integer numFluxSupportPoints >
+  GEOS_HOST_DEVICE
+  static void
+  compute ( integer const numPhase,
+            integer const ip,
+            integer const hasCapPressure,
+            localIndex const ( &seri )[numFluxSupportPoints],
+            localIndex const ( &sesri )[numFluxSupportPoints],
+            localIndex const ( &sei )[numFluxSupportPoints],
+            real64 const ( &trans )[numFluxSupportPoints],
+            real64 const ( &dTrans_dPres )[numFluxSupportPoints],
+            ElementViewConst< arrayView1d< real64 const > > const & pres,
+            ElementViewConst< arrayView1d< real64 const > > const & gravCoef,
+            ElementViewConst< arrayView3d< real64 const, compflow::USD_PHASE_DC > > const & dPhaseVolFrac,
+            ElementViewConst< arrayView3d< real64 const, constitutive::multifluid::USD_PHASE > > const & phaseMassDens,
+            ElementViewConst< arrayView4d< real64 const, constitutive::multifluid::USD_PHASE_DC > > const & dPhaseMassDens,
+            ElementViewConst< arrayView3d< real64 const, constitutive::cappres::USD_CAPPRES > > const & phaseCapPressure,
+            ElementViewConst< arrayView4d< real64 const, constitutive::cappres::USD_CAPPRES_DS > > const & dPhaseCapPressure_dPhaseVolFrac,
+            real64 & potGrad,
+            real64 ( & dPresGrad_dP )[numFluxSupportPoints],
+            real64 ( & dPresGrad_dC )[numFluxSupportPoints][numComp],
+            real64 ( & dGravHead_dP )[numFluxSupportPoints],
+            real64 ( & dGravHead_dC )[numFluxSupportPoints][numComp] )
+  {
+    // assign derivatives arrays to zero
+    for( integer i = 0; i < numFluxSupportPoints; ++i )
+    {
+      dPresGrad_dP[i] = 0.0;
+      dGravHead_dP[i] = 0.0;
+      for( integer jc = 0; jc < numComp; ++jc )
+      {
+        dPresGrad_dC[i][jc] = 0.0;
+        dGravHead_dC[i][jc] = 0.0;
+      }
+    }
+
+    // create local work arrays
+    real64 densMean = 0.0;
+    real64 dDensMean_dP[numFluxSupportPoints]{};
+    real64 dDensMean_dC[numFluxSupportPoints][numComp]{};
+
+    real64 presGrad = 0.0;
+    real64 gravHead = 0.0;
+    real64 dCapPressure_dC[numComp]{};
+
+    // calculate quantities on primary connected cells
+    for( integer i = 0; i < numFluxSupportPoints; ++i )
+    {
+      localIndex const er  = seri[i];
+      localIndex const esr = sesri[i];
+      localIndex const ei  = sei[i];
+
+      // density
+      real64 const density  = phaseMassDens[er][esr][ei][0][ip];
+      real64 const dDens_dP = dPhaseMassDens[er][esr][ei][0][ip][Deriv::dP];
+
+      // average density and derivatives
+      densMean += 0.5 * density;
```

Comment:
this needs to be synced with current treatment

Reply:
Thanks for pointing this out. Synced.